tools/man/hio_ds_info.1
.\" -*- nroff -*-
.\" Copyright 2016 Los Alamos National Security, LLC. All rights
.\" reserved.
.\" $COPYRIGHT$
.TH hio_ds_info 1 "October 2016" libhio "User Manuals"
.SH NAME
hio_ds_info \- dump hio dataset metadata
.SH SYNOPSIS
\fBhio_ds_info [--all] [--elements] [--rank <rank>] [--config]
[--perform] <data root> <context name> [dataset name [dataset id]]\fP
.SH DESCRIPTION
\fBhio_ds_info\fP dumps the metadata from dataset(s) in a
human-readable format.
.SH OPTIONS
.IP --all | -a
Dump all the metadata from dataset(s).
.IP --elements | -e
Dump element metadata from dataset(s).
.IP --rank | -r
Dump metadata associated with a specific rank (unique offset mode
only).
.IP --config | -c
Dump dataset configuration variables.
.IP --perform | -p
Dump dataset performance variables.
.SH AUTHOR
Nathan Hjelm <hjelmn at cs dot unm dot edu>
\fB


NEWS
HIO NEWS
========

Last updated 2016-08-14

Release hio.1.3.0.6

Cherry pick several test enhancements from 1.4 back into 1.3.  Timestamp more
run_setup messages, gather info with lstopo, add f* file operations, better
cpuinfo listing, add ifr/eif actions, add signal handling.

Release hio.1.3.0.5

This release adds additional performance instrumentation and diagnostic
messages on test failure.  A new action "dbuf" is added to xexec to specify
the write data pattern used for data compare. The test cases have been updated
to specify the pattern previously used by default. 

Release hio.1.3.0.4

Multiple test enhancements, better messages on certain failures, job specific
DataWarp information logged.

Release hio.1.3.0.3

Some small test improvements, a fix for partial read and write handling and
improved diagnostic and debug messages related to dw_stage_directory_out calls.
Additionally, an early version of an HDF5 plugin has been added. It is untested
and is not included in the distribution tarball.

Release hio.1.3.0.2

A few minor updates, all related to testing.

Release hio.1.3.0.1

Fixes for getting incorrect stripe count, some logging related improvements
and performance enhancements.  runxx and run_combo tests now support
-k (processor type) and -u (debugger) options. 

Release hio.1.3.0.0

This is the first open source relase of HIO.  It is now available on github
at https://github.com/hpc/libhio .  We're still working out our release
procedures, so for now the release is designated by the tag hio.1.3.0.0.

Changes since the rc version are primarily fixes and performance enhancements 
to optimised mode.  Additionally, there are a number of minor test fixes and
functional enhancements.

Release hio.1.3.0.0-rc1:

Updated external interface to allow for configuration of datasets before
open. Two existing functions have been modified and two new functions have
been added. The new functions are hio_dataset_alloc and hio_dataset_free
which take the same arguments as the old hio_dataset_open and
hio_dataset_close respectively. hio_dataset_open and hio_dataset_close now
only take a hio_dataset_t handle. After a dataset has been allocated using
hio_dataset_alloc it can be configured using hio_config_set_value. A new
function hio_object_get_name() has been added to get the name associated
with a context, dataset, or element handle.  The example in the API document
has not yet been updated.  See tests/test01.c for a working example of the
modified API.  Numerous minor documentation and test fixes and enhancements.

Release hio.1.2.0.4:

Make hio.h C89 compliant.  Minor build and test updates.

Release hio.1.2.0.3:

Minor documentation and test fixes and enhancements.


Release hio.1.2.0.2:

Updated statistics reported when HIO_print_statistics is set to report the
overall (dataset open -> close) read/write speed and instantaneous
read/write speed. The speed is now only reported by the first rank in the
context. Removed error messages when 0-byte dataset manifests are
encountered. Fixed bugs when using datawarp including deleting the wrong
dataset ID after staging a new dataset ID. Added new variables to expose the
RAID level and filesystem type backing a dataset. New manifests are now
written ending in a newline instead of a NULL character. This does not
break backwards compatibility with 1.2.x.x releases.  Added staging test
run20.  Added Trinity step-by-step build and run procedure to README.
Minor build and test improvements.


Release hio.1.2.0.1:

Test scripts enhanced with new, larger size options y and z. Multi-test
script run_combo added.  Added -m <msub_args> option. Also, additional
minor build and testcase fixes.


Release hio.1.2.0.0:

Switched from XML to JSON format for the manifest file.  This means that hio
datasets created with previous releases cannot be read with this version. We
don't expect to be this casual about incompatible changes in the future, but
since at the moment, hio is only being used for development and testing
activities an exception seems reasonable.

Minor bug fixes, diagnostic, testing and build improvements.

--- End of NEWS ---


tools/hio_ds_info.c
/* -*- Mode: C; c-basic-offset:2 ; indent-tabs-mode:nil -*- */
/*
 * Copyright (c) 2016-2017 Los Alamos National Security, LLC.  All rights
 *                         reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 */

#include <stdlib.h>
#include <getopt.h>
#include <errno.h>
#include <string.h>

#include "hio.h"
#include "hio_internal.h"

static void print_usage (void) {
  printf ("Usage: hio_ds_info [options] <data root> <context name> [dataset name [dataset id]]\n");
  printf ("Dump metadata associated with libhio dataset(s).\n");

  exit (EXIT_FAILURE);
}

static void print_version (void) {
  printf ("hio_ds_info %s", VERSION);
  printf ("Copyright © 2016 Los Alamos National Security, LLC\n\n");

  printf("libhio comes with NO WARRANTY.\n");
  printf("You may redistribute copies of libhio under the terms\n");
  printf("of the BSD 3-Clause License.\n");
  printf("For more information about these issues\n");
  printf("see the file named COPYING in the libhio distribution.\n");

  exit (EXIT_SUCCESS);
}

int main (int argc, char *argv[]) {
  const char *data_root, *context = NULL, *dataset = NULL;
  int rc, opt, opt_index, list_rank = -1;
  int64_t dataset_id = -1;
  uint32_t flags = 0;

  struct option long_options[] = {
    {.name = "all", .has_arg = no_argument, .flag = NULL, .val = 'a'},
    {.name = "config", .has_arg = no_argument, .flag = NULL, .val = 'c'},
    {.name = "elements", .has_arg = no_argument, .flag = NULL, .val = 'e'},
    {.name = "help", .has_arg = no_argument, .flag = NULL, .val = 'h'},
    {.name = "perform", .has_arg = no_argument, .flag = NULL, .val = 'p'},
    {.name = "rank", .has_arg = required_argument, .flag = NULL, .val = 'r'},
    {.name = "version", .has_arg = no_argument, .flag = NULL, .val = 'v'},
  };

  do {
    opt = getopt_long (argc, argv, "aer:hvcp", long_options, &opt_index);
    if (-1 == opt) {
      break;
    }

    switch (opt) {
    case 'a':
      flags = HIO_DUMP_FLAG_MASK;
      break;
    case 'c':
      flags |= HIO_DUMP_FLAG_CONFIG;
      break;
    case 'e':
      flags |= HIO_DUMP_FLAG_ELEMENTS;
      break;
    case 'h':
      print_usage ();
      break;
    case 'p':
      flags |= HIO_DUMP_FLAG_PERF;
      break;
    case 'r':
      list_rank = strtol (optarg, NULL, 0);
      break;
    case 'v':
      print_version ();
      break;
    default:
      fprintf (stderr, "Invalid option \'%c\' specified.\n", opt);
      print_usage ();
    }
  } while (1);

  if (argc - optind < 2) {
    print_usage ();
  }

  data_root = argv[optind++];
  context = argv[optind++];
  if (argc - optind) {
    dataset = argv[optind++];
    if (argc - optind) {
      errno = 0;
      dataset_id = strtol (argv[optind], NULL, 10);
      if (0 != errno) {
        if (0 == strcasecmp (argv[optind], "newest")) {
          dataset_id = HIO_DATASET_ID_NEWEST;
        } else if (0 == strcasecmp (argv[optind], "highest_id")) {
          dataset_id = HIO_DATASET_ID_HIGHEST;
        } else {
          fprintf (stderr, "Invalid dataset id: %s. Valid values: positive integer, newest, highest_id\n", argv[optind]);
        }
      }
    }
  }

  hio_dataset_dump (data_root, context, dataset, dataset_id, flags, list_rank, stdout);

  return 0;
}
